cdap-app-fabric/src/test/java/co/cask/cdap/internal/audit/AuditPublishTest.java

```diff
@@ -0,0 +1,130 @@
+
+/**
+ * Tests audit publishing.
+ */
+public class AuditPublishTest {
+  @ClassRule
+  public static final KafkaTester KAFKA_TESTER = new KafkaTester(
+    ImmutableMap.of(Constants.Audit.ENABLED, "true"), Collections.<Module>emptyList(), 1);
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(AuditMessage.class, new AuditMessageTypeAdapter())
+    .registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())
+    .create();
+
+  private static ZKClientService zkClient;
+  private static KafkaClientService kafkaClient;
+
+  @BeforeClass
+  public static void init() throws Exception {
+    Injector injector = AppFabricTestHelper.getInjector(KAFKA_TESTER.getCConf(),
+                                                        new AbstractModule() {
+                                                          @Override
+                                                          protected void configure() {
+                                                            install(new AuditModule().getDistributedModules());
+                                                            install(new ZKClientModule());
+                                                            install(new KafkaClientModule());
+                                                          }
+                                                        });
+
+    zkClient = injector.getInstance(ZKClientService.class);
+    zkClient.startAndWait();
+    kafkaClient = injector.getInstance(KafkaClientService.class);
+    kafkaClient.startAndWait();
+  }
+
+  @AfterClass
+  public static void stop() throws Exception {
+    zkClient.stopAndWait();
+    kafkaClient.startAndWait();
+  }
+
+  @Test
+  public void testPublish() throws Exception {
+    String defaultNs = NamespaceId.DEFAULT.getNamespace();
+    String appName = WordCountApp.class.getSimpleName();
+    Set<? extends EntityId> expectedMetadataChangeEntities =
+      ImmutableSet.of(Ids.namespace(defaultNs).artifact("app", "1"),
+                      Ids.namespace(defaultNs).app(appName),
+                      Ids.namespace(defaultNs).app(appName).flow(WordCountApp.WordCountFlow.class.getSimpleName()),
+                      Ids.namespace(defaultNs).app(appName).mr(WordCountApp.VoidMapReduceJob.class.getSimpleName()),
+                      Ids.namespace(defaultNs).app(appName)
+                        .service(WordCountApp.WordFrequencyService.class.getSimpleName()),
+                      Ids.namespace(defaultNs).dataset("mydataset"),
+                      Ids.namespace(defaultNs).stream("text"));
+
+    Multimap<AuditType, EntityId> expectedAuditEntities = HashMultimap.create();
+    expectedAuditEntities.putAll(AuditType.METADATA_CHANGE, expectedMetadataChangeEntities);
+
+    AppFabricTestHelper.deployApplication(WordCountApp.class);
+    List<AuditMessage> publishedMessages =
+      KAFKA_TESTER.getPublishedMessages(KAFKA_TESTER.getCConf().get(Constants.Audit.KAFKA_TOPIC), 11,
```

Comment:
We should have waitFor for this. I've seen similar test case failure without using waitFor.